src/installer/managed/Microsoft.NET.HostModel/MachO/MachObjectFile.cs

```diff
@@ -0,0 +1,500 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.IO.MemoryMappedFiles;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Microsoft.NET.HostModel.MachO;
+
+/// <summary>
+/// A managed object containing relevant information for AdHoc signing a Mach-O file.
+/// The object is created from a memory mapped file, and a signature can be calculated from the memory mapped file.
+/// However, since a memory mapped file cannot be extended, the signature is written to a file stream.
+/// </summary>
+internal class MachObjectFile
+{
+    internal const uint SpecialSlotCount = 2;
+    internal const uint PageSize = 4096;
+    internal const byte Log2PageSize = 12;
+    internal const byte DefaultHashSize = 32;
+    internal const HashType DefaultHashType = HashType.SHA256;
+    internal static IncrementalHash GetDefaultIncrementalHash() => IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
+
+    private MachHeader _header;
+    private (LinkEditCommand Command, long FileOffset) _codeSignatureLC;
+    private readonly (SegmentLoadCommand Command, long FileOffset) _textSegment;
+    private readonly (Segment64LoadCommand Command, long FileOffset) _textSegment64;
+    private (SegmentLoadCommand Command, long FileOffset) _linkEditSegment;
+    private (Segment64LoadCommand Command, long FileOffset) _linkEditSegment64;
+    private readonly long _lowestSection;
+    private readonly string _identifier;
+    private CodeSignature _codeSignatureBlob;
+    private readonly long _nextCommandPtr;
+
+    /// <summary>
+    /// Creates a new MachObjectFile from a memory mapped file.
+    /// </summary>
+    public MachObjectFile(MemoryMappedViewAccessor file, string identifier)
+    {
+        long commandsPtr = 0;
+        file.Read(commandsPtr, out _header);
+        _nextCommandPtr = ReadCommands(file, in _header, out _codeSignatureLC, out _textSegment, out _textSegment64, out _linkEditSegment, out _linkEditSegment64, out _lowestSection);
```

Comment:
This is a lot of work to do in a constructor. If you're going to read things out of the file, I would consider inverting the control flow and having a static method that reads things out of the file and then inject the values into the constructor.